packages/plugin-react/src/index.ts

```diff
@@ -181,114 +185,128 @@ export default function viteReact(opts: Options = {}): PluginOption[] {
         // we only create static option in this case and re-create them
         // each time otherwise
         staticBabelOptions = createBabelOptions(opts.babel)
+
+        if (
+          canSkipBabel(staticBabelOptions.plugins, staticBabelOptions) &&
+          skipFastRefresh &&
+          isProduction
+        ) {
+          delete viteBabel.transform
+        }
       }
     },
-    async transform(code, id, options) {
-      if (id.includes('/node_modules/')) return
-
-      const [filepath] = id.split('?')
-      if (!filter(filepath)) return
-
-      const ssr = options?.ssr === true
-      const babelOptions = (() => {
-        if (staticBabelOptions) return staticBabelOptions
-        const newBabelOptions = createBabelOptions(
-          typeof opts.babel === 'function'
-            ? opts.babel(id, { ssr })
-            : opts.babel,
-        )
-        runPluginOverrides?.(newBabelOptions, { id, ssr })
-        return newBabelOptions
-      })()
-      const plugins = [...babelOptions.plugins]
-
-      const isJSX = filepath.endsWith('x')
-      const useFastRefresh =
-        !skipFastRefresh &&
-        !ssr &&
-        (isJSX ||
-          (opts.jsxRuntime === 'classic'
-            ? importReactRE.test(code)
-            : code.includes(jsxImportDevRuntime) ||
-              code.includes(jsxImportRuntime)))
-      if (useFastRefresh) {
-        plugins.push([
-          await loadPlugin('react-refresh/babel'),
-          { skipEnvCheck: true },
-        ])
-      }
-
-      if (opts.jsxRuntime === 'classic' && isJSX) {
-        if (!isProduction) {
-          // These development plugins are only needed for the classic runtime.
-          plugins.push(
-            await loadPlugin('@babel/plugin-transform-react-jsx-self'),
-            await loadPlugin('@babel/plugin-transform-react-jsx-source'),
+    transform: {
+      filter: {
+        id: {
+          include: ensureArray(include).map(matchWithQuery),
+          exclude: [
+            ...(exclude ? ensureArray(exclude).map(matchWithQuery) : []),
+            /\/node_modules\//,
+          ],
```

Comment:
This filter conversion and the two utils feels a bit tricky IMO.
Could this be upstream into the Vite `createFilter` function so that it can be reused by other plugins?

Reply:
If we add a util function to Vite, we need to bump the required peer dep in this plugin (and the other plugins that wants to use that) which is a breaking change. I think if we do that we probably should add it to a separate package.

I'll consider if we can have `@rolldown/pluginutils` (since I sent a PR to `@rollup/pluginutils` that adds some utils functions and it's taking some time to be merged).

Reply:
This could be added and then we could do `import * as Vite from 'vite'` and check for the export name existence so that people using the latest version have the speedup (which is expecting for people running rolldown vite)

Reply:
I've updated the code to use `@rolldown/pluginutils`. `ensureArray` is still there because we need to combine with `/\/node_modules\//`, but I think it is now simple enough.